utilities/env_librados.cc

```diff
@@ -0,0 +1,1043 @@
+const int MAX_ITEMS_IN_FS = 1<<20;
+const std::string TEST_DIR_PREFIX = "TEST_TEMP_";
+const std::string ROOT_DIR_TAG = "/";
+
+rocksdb::Status err_to_status(int r)
+{
+  switch (r) {
+  case 0:
+    return rocksdb::Status::OK();
+  case -ENOENT:
+  case -ENODATA:
+  case -ENOTDIR:
+    return rocksdb::Status::NotFound(rocksdb::Status::kNone);
+  case -EINVAL:
+    return rocksdb::Status::InvalidArgument(rocksdb::Status::kNone);
+  case -EIO:
+    return rocksdb::Status::IOError(rocksdb::Status::kNone);
+  default:
+    // FIXME :(
+    assert(0 == "unrecognized error code");
+    return rocksdb::Status::NotSupported(rocksdb::Status::kNone);
+  }
+}
+
+void split(const std::string &fn, std::string *dir, std::string *file) {
+  size_t slash = fn.rfind('/');
+  *file = fn.substr(slash + 1);
+  while (slash && fn[slash-1] == '/')
+    --slash;
+  *dir = fn.substr(0, slash);
+
+  if (*dir == "") {
+      *dir = ROOT_DIR_TAG;
+  }
+}
+
+// A file abstraction for reading sequentially through a file
+class LibradosSequentialFile : public rocksdb::SequentialFile {
+    librados::IoCtx _io_ctx;
+    std::string _fid;
+    int _offset;
+ public:
+  LibradosSequentialFile(librados::IoCtx io_ctx, std::string fid):
+    _io_ctx(io_ctx), _fid(fid), _offset(0){}
+  ~LibradosSequentialFile() {}
+
+  // Read up to "n" bytes from the file.  "scratch[0..n-1]" may be
+  // written by this routine.  Sets "*result" to the data that was
+  // read (including if fewer than "n" bytes were successfully read).
+  // May set "*result" to point at data in "scratch[0..n-1]", so
+  // "scratch[0..n-1]" must be live when "*result" is used.
+  // If an error was encountered, returns a non-OK status.
+  //
+  // REQUIRES: External synchronization
+  rocksdb::Status Read(size_t n, rocksdb::Slice* result, char* scratch) {
+    librados::bufferlist buffer;
+    int r = _io_ctx.read(_fid, buffer, n, _offset);
+    if (r >= 0) {
+        buffer.copy(0, r, scratch);
+        LOG_DEBUG("%i\n", r);
+        *result = rocksdb::Slice(scratch, r);
+        _offset += r;
+        return rocksdb::Status::OK();
+    } else {
+        return err_to_status(r);
+    }
+  }
+
+  // Skip "n" bytes from the file. This is guaranteed to be no
+  // slower that reading the same data, but may be faster.
+  //
+  // If end of file is reached, skipping will stop at the end of the
+  // file, and Skip will return OK.
+  //
+  // REQUIRES: External synchronization
+  rocksdb::Status Skip(uint64_t n) {
+    _offset += n;
+    return rocksdb::Status::OK();
+  }
+
+  // Remove any kind of caching of data from the offset to offset+length
+  // of this file. If the length is 0, then it refers to the end of file.
+  // If the system is not caching the file contents, then this is a noop.
+  rocksdb::Status InvalidateCache(size_t offset, size_t length) {
+  #if 0
+    fs->invalidate_cache(h->file, offset, length);
+  #endif
+    /* not caching, do nothing */
+    return rocksdb::Status::OK();
```

Comment:
rocksdb has it's own caching capabilities that we should be able to use, without relying on a cache here.  This can safely be a no-op.